src/main/java/org/junitpioneer/jupiter/SystemPropertyExtension.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2015-2020 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junitpioneer.jupiter;
+
+import static java.util.stream.Collectors.*;
+
+import java.lang.annotation.Annotation;
+import java.util.*;
+import java.util.stream.Stream;
+
+import org.junit.jupiter.api.extension.AfterAllCallback;
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionConfigurationException;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ExtensionContext.Namespace;
+import org.junit.platform.commons.support.AnnotationSupport;
+
+class SystemPropertyExtension implements BeforeAllCallback, BeforeEachCallback, AfterAllCallback, AfterEachCallback {
+
+	private static final Namespace NAMESPACE = Namespace.create(SystemPropertyExtension.class);
+	private static final String BACKUP = "Backup";
+
+	@Override
+	public void beforeAll(ExtensionContext context) throws Exception {
+		handleSystemProperties(context);
+	}
+
+	@Override
+	public void beforeEach(ExtensionContext context) throws Exception {
+		if (annotationsPresentOnTestMethod(context)) {
+			handleSystemProperties(context);
+		}
+	}
+
+	private boolean annotationsPresentOnTestMethod(ExtensionContext context) {
+		//@formatter:off
+		return context.getTestMethod()
+				.map(testMethod -> AnnotationSupport.isAnnotated(testMethod, ClearSystemProperty.class)
+						|| AnnotationSupport.isAnnotated(testMethod, ClearSystemProperties.class)
+						|| AnnotationSupport.isAnnotated(testMethod, SetSystemProperty.class)
+						|| AnnotationSupport.isAnnotated(testMethod, SetSystemProperties.class))
+				.orElse(false);
+		//@formatter:on
+	}
+
+	private void handleSystemProperties(ExtensionContext context) {
+		Set<String> propertiesToClear;
+		Map<String, String> propertiesToSet;
+		try {
+			propertiesToClear = findRepeatableAnnotations(context, ClearSystemProperty.class).stream()
+					// collect to map because the collector throws an IllegalStateException on
```

Comment:
Generally: if there is a need for an inline comment then it is a code smell. In this case, the code uses a `Map` instead of a `Set` because the collector to the `Map` throws an exception, while the collector to the `Set` does not (which is btw. absolutely logical).

The nice solution is to use a special set that throws an exception when an element already in the set is added:

```java
private static class UniqueThrowingSet extends HashSet {
        @Override
        public boolean add(Object o) {
            if (contains(o)) {
                throw new IllegalStateException("Duplicate element cannot be added");
            }
            return super.add(o);
        }
    }
```
I am not advocating this solution against the current solution, because this seems to be a bit overkill for the actual problem.